File: hacker-rank/thirty-days-of-code/day-20/solution.js
'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', function(inputStdin) {
    inputString += inputStdin;
});

process.stdin.on('end', function() {
    inputString = inputString.split('\n');

    main();
});

function readLine() {
    return inputString[currentLine++];
}



function main() {
    const n = parseInt(readLine().trim(), 10);

    const a = readLine().replace(/\s+$/g, '').split(' ').map(aTemp => parseInt(aTemp, 10));

    // Write your code here
    let totalSwaps = 0;

    for (let i = 0; i < n; i++) {
      // Track number of elements swapped during a single array traversal
      let swapsThisPass = 0;
      
      for (let j = 0; j < n - 1; j++) {
        // Swap adjacent elements if they are in decreasing order
        let current = a[j];
        let next = a[j + 1];
        if (current > next) {
          a[j] = next;
          a[j + 1] = current;
          swapsThisPass++;
        }
      }
      totalSwaps = totalSwaps + swapsThisPass;
      // If no elements were swapped during a traversal, array is sorted
      if (swapsThisPass == 0) {
        break;
      }
    }

    console.log("Array is sorted in " + totalSwaps + " swaps.");
    console.log("First Element: " + a[0]);
    console.log("Last Element: " + a[a.length - 1]);
    
}

File: hacker-rank/thirty-days-of-code/day-20/Solution.java
import java.io.*;
import java.math.*;
import java.security.*;
import java.text.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;
import java.util.regex.*;
import java.util.stream.*;
import static java.util.stream.Collectors.joining;
import static java.util.stream.Collectors.toList;



public class Solution {

  public static void main(String[] args) throws IOException {
    BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(System.in));

    int n = Integer.parseInt(bufferedReader.readLine().trim());

    List<Integer> a = Stream.of(bufferedReader.readLine().replaceAll("\\s+$", "").split(" "))
    .map(Integer::parseInt)
    .collect(toList());

    // Write your code here
    int totalSwaps = 0;

    for (int i = 0; i < n; i++) {
      // Track number of elements swapped during a single array traversal
      int swapsThisPass = 0;
      
      for (int j = 0; j < n - 1; j++) {
        // Swap adjacent elements if they are in decreasing order
        int current = a.get(j);
        int next = a.get(j + 1);
        if (current > next) {
          a.set(j, next);
          a.set(j + 1, current);
          swapsThisPass++;
        }
      }
      totalSwaps = totalSwaps + swapsThisPass;
      // If no elements were swapped during a traversal, array is sorted
      if (swapsThisPass == 0) {
        break;
      }
    }

    System.out.println("Array is sorted in " + totalSwaps + " swaps.");
    System.out.println("First Element: " + a.get(0));
    System.out.println("Last Element: " + a.get(a.size() - 1));
    
    bufferedReader.close();
  }
}
